client/go/outline/platerrors/platform_error.go

```diff
@@ -23,89 +23,73 @@ import (
-func NewWithDetails(code ErrorCode, message string, details ErrorDetails) *PlatformError {
-	return &PlatformError{
-		Code:    code,
-		Message: message,
-		Details: details,
+// ToPlatformError converts an [error] into a [PlatformError].
+// If the provided err is already a [PlatformError], it is returned as is.
+// Otherwise, the err is wrapped into a new [PlatformError] of [InternalError].
+// It returns nil if err is nil.
+func ToPlatformError(err error) *PlatformError {
+	if err == nil {
+		return nil
 	}
-}
-
-// NewWithCause creates a [PlatformError] with a specific error code, message, and cause.
-func NewWithCause(code ErrorCode, message string, cause error) *PlatformError {
-	return &PlatformError{
-		Code:    code,
-		Message: message,
-		Cause:   cause,
+	if pe, ok := err.(PlatformError); ok {
+		return &pe
 	}
-}
-
-// NewWithDetailsCause creates a [PlatformError] with an error code, message, details, and cause.
-func NewWithDetailsCause(code ErrorCode, message string, details ErrorDetails, cause error) *PlatformError {
-	return &PlatformError{
-		Code:    code,
-		Message: message,
-		Details: details,
-		Cause:   cause,
+	if pe, ok := err.(*PlatformError); ok {
+		return pe
 	}
+	return &PlatformError{Code: InternalError, Message: err.Error()}
 }
 
 // Error returns a JSON string containing the error details and all its underlying causes,
 // until it finds a cause that is not a [PlatformError].
 // The resulting JSON can be used to reconstruct the error in TypeScript.
-func (e *PlatformError) Error() string {
-	if e == nil {
-		return formatInternalErrorJSON("nil error")
-	}
-	if e.Code == "" {
-		return formatInternalErrorJSON("empty error code")
+func (e PlatformError) Error() string {
+	e.normalize()
+	msg := fmt.Sprintf("(%v) %v", e.Code, e.Message)
+	if e.Cause != nil {
+		msg += fmt.Sprintf(": %v", e.Cause)
 	}
-	errJson, err := json.Marshal(convertToPlatformErrJSON(e))
-	if err != nil {
-		return formatInternalErrorJSON("JSON marshal failure: " + err.Error())
-	}
-	return string(errJson)
+	return msg
 }
 
 // Unwrap returns the cause of this [PlatformError].
-func (e *PlatformError) Unwrap() error {
+func (e PlatformError) Unwrap() error {
 	return e.Cause
 }
 
-// formatInternalErrorJSON creates a JSON string with ErrorCode InternalError and the msg.
-// Please note that msg should not contain double quotes.
-// The returned JSON string is compatible with the JSON returned by [PlatformError].Error().
-func formatInternalErrorJSON(msg string) string {
-	rmEscapes := strings.NewReplacer("\\", "", "\"", "")
-	return fmt.Sprintf(
-		`{"code":"%s","message":"%s"}`,
-		InternalError, rmEscapes.Replace(msg))
+// MarshalJSONString returns a JSON string containing the error details and
+// all its underlying causes.
+// The resulting JSON can be used to reconstruct the error in TypeScript.
+//
+// This function guarantees to always return a non-empty JSON string.
+func (e PlatformError) MarshalJSONString() string {
```

Comment:
Please move the MarshalJSONString out of the type. There's no need to be part of the type, since there's no private member.  Understanding things in isolation reduces complexity.

Also, this is marshaling specific to an IPC mechanism, it should be something that could be moved to another library if we wanted to.

Reply:
Make it return an error in case of failed to marshall, so the caller can decide what to do.

Reply:
Sounds good, we may construct a "failed to marshal error" in Java/Swift.